src/local-state/__tests__/LocalState/base.test.ts

```diff
@@ -0,0 +1,1103 @@
+
+  localState.addResolvers({
+    Query: {
+      foo: () => ({ __typename: "Foo", bar: true }),
+    },
+  });
+
+  await expect(
+    localState.execute({
+      document,
+      client,
+      context: {},
+      variables: {},
+      remoteResult: undefined,
+    })
+  ).resolves.toStrictEqualTyped({
+    data: { foo: { __typename: "Foo", bar: true } },
+  });
+});
+
+test("handles queries with a mix of @client and server fields", async () => {
+  const document = gql`
+    query Mixed {
+      foo @client {
+        bar
+      }
+      bar {
+        baz
+      }
+    }
+  `;
+
+  const client = new ApolloClient({
+    cache: new InMemoryCache(),
+    link: ApolloLink.empty(),
+  });
+  const remoteResult = { data: { bar: { __typename: "Bar", baz: true } } };
+
+  const localState = new LocalState({
+    resolvers: {
+      Query: {
+        foo: () => ({ __typename: "Foo", bar: true }),
+      },
+    },
+  });
+
+  await expect(
+    localState.execute({
+      document,
+      client,
+      context: {},
+      variables: {},
+      remoteResult,
+    })
+  ).resolves.toStrictEqualTyped({
+    data: {
+      foo: { __typename: "Foo", bar: true },
+      bar: { __typename: "Bar", baz: true },
+    },
+  });
+});
+
+test("runs resolvers for deeply nested @client fields", async () => {
+  const document = gql`
+    query Test {
+      user {
+        id
+        bestFriend {
+          firstName
+          lastName
+          fullName @client
+        }
+      }
+    }
+  `;
+
+  const client = new ApolloClient({
+    cache: new InMemoryCache(),
+    link: ApolloLink.empty(),
+  });
+
+  const remoteResult = {
+    data: {
+      user: {
+        __typename: "User",
+        id: 1,
+        bestFriend: {
+          __typename: "User",
+          firstName: "Test",
+          lastName: "User",
+        },
+      },
+    },
+  };
+
+  const localState = new LocalState({
+    resolvers: {
+      User: {
+        fullName: (user) => `${user.firstName} ${user.lastName}`,
+      },
+    },
+  });
+
+  await expect(
+    localState.execute({
+      document,
+      client,
+      context: {},
+      variables: {},
+      remoteResult,
+    })
+  ).resolves.toStrictEqualTyped({
+    data: {
+      user: {
+        __typename: "User",
+        id: 1,
+        bestFriend: {
+          __typename: "User",
+          firstName: "Test",
+          lastName: "User",
+          fullName: "Test User",
+        },
+      },
+    },
+  });
+});
+
+test("has access to query variables in @client resolvers", async () => {
+  const document = gql`
+    query WithVariables($id: ID!) {
+      foo @client {
+        bar(id: $id)
+      }
+    }
+  `;
+
+  const client = new ApolloClient({
+    cache: new InMemoryCache(),
+    link: ApolloLink.empty(),
+  });
+
+  const localState = new LocalState({
+    resolvers: {
+      Query: {
+        foo: () => ({ __typename: "Foo" }),
+      },
+      Foo: {
+        bar: (_: any, { id }: { id: number }) => id,
```

Comment:
Let's use different names for the parameter and variable so we can be sure that this is not the variable being passed through directly, but a field parameter that is taking on the value of the variable.


```suggestion
        bar(fromVariable: $id)
      }
    }
  `;

  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: ApolloLink.empty(),
  });

  const localState = new LocalState({
    resolvers: {
      Query: {
        foo: () => ({ __typename: "Foo" }),
      },
      Foo: {
        bar: (_: any, { fromVariable }: { fromVariable: number }) =>
          fromVariable,
```

Reply:
Updated in [0266052](https://github.com/apollographql/apollo-client/pull/12614/commits/0266052407a7469d3a7c0142d4c09a8616ba8e95). I went ahead and removed the type casting for now since it isn't doing much in this test (especially now that we have better values for resolver functions out of the box)